### ISSUE TITLE
Incorrect inference with ambiguous overloads with mutiple arguments

### ISSUE THREAD

**Issue by randolf-scholz** [astral-sh/ty#2906]
### Summary

https://play.ty.dev/7e986f4b-4688-4a6f-951b-f78ef9eebb09

```python
from typing import Any, overload, assert_type

class A[T]:  # covariant
    def get(self) -> T: ...

@overload
def op(l: A[None], r: A[None]) -> A[None]: ...
@overload
def op(l: A[None], r: A[Any]) -> A[None]: ...
@overload
def op(l: A[Any], r: A[None]) -> A[None]: ...
@overload
def op(l: A[Any], r: A[Any]) -> A[Any]: ...

def test(x: A[None], y: A[Any]) -> None:
    assert_type(op(x, x), A[None])  # spec: ✅️ 
    assert_type(op(x, y), A[None])  # spec: ✅️
    assert_type(op(y, x), A[None])  # spec: ✅️
    assert_type(op(y, y), Any)     # spec: ❌️ (expected Any)
```

`ty` incorrectly infers `A[None]` for `op(A[Any], A[Any])`. See https://github.com/python/typing/issues/2196 and https://github.com/astral-sh/ty/issues/665 for context.

Single argument ambiguous overloads seem to currently infer as `Unknown`.

**Comment by carljm** [comment]
Thanks for the report! It looks like pyright and pyrefly share the same behavior as ty here, while mypy/zuban infer `A[Any]` for the last three `assert_type`. So it seems no type checker is quite in compliance with the spec here currently.
